### DISCUSSION TITLE
Trying to use VASP recipe for the first time

### DISCUSSION THREAD

**Discussion by JohnFengg** [Quantum-Accelerators/quacc#1038] [General]
I'm trying to using QUACC package with JOBFLOW wf engine to run a job, my script is like this (got an error before testing if this scripts can work):
####extract the atom objects from database ####
from ase.db import connect
from quacc.recipes.vasp.mp import mp_relax_flow
from quacc.wflow.db import results_to_db

database_path = 'perovskites.db'
db = connect(database_path)

batch_size = 1 #number of structure calculated simultaneously

for i in range(0, 1, batch_size):
structures = db[i:i + batch_size]
flow=mp_relax_flow()
#pre_relax={atoms=structures, preset="MPScanSet", calc_swaps= None, copy_files=None}
results = flow(
structures,
prerelax_job_kwargs=None,
relax_job_kwargs=None,
)

db.close()

when I run the scripts, there is an error:

Traceback (most recent call last):
File "/home/toussaint/high-throughput/model-run.py", line 3, in
from quacc.recipes.vasp.mp import mp_relax_flow
File "/home/toussaint/opt/miniconda3/envs/quacc/lib/python3.10/site-packages/quacc/init.py", line 11, in
from quacc.settings import QuaccSettings
File "/home/toussaint/opt/miniconda3/envs/quacc/lib/python3.10/site-packages/quacc/settings.py", line 22, in
import jobflow
File "/home/toussaint/opt/miniconda3/envs/quacc/lib/python3.10/site-packages/jobflow/init.py", line 14, in
SETTINGS = JobflowSettings()
File "pydantic/env_settings.py", line 40, in pydantic.env_settings.BaseSettings.init
File "pydantic/main.py", line 341, in pydantic.main.BaseModel.init
pydantic.error_wrappers.ValidationError: 1 validation error for JobflowSettings
docs_store
extra fields not permitted (type=value_error.extra)

I have created a jobflow.yaml file and export the path in ~/.bashrc, and it seems like there are no error about the yaml files:

docs_store:
type: MongoStore
host: localhost
port: 27017
database: ase
collection_name: ase
username: toussaint
password: toussaint

I dont know how to fix this, could anyone help me for this?

**Comment by Andrew-S-Rosen** [comment]
Hi @JohnFengg. Thanks for the report. I'd be happy to help. 

The error you are seeing is likely a Pydantic compatibility issue. In short, the entire Materials Project software stack ([including Jobflow](https://github.com/materialsproject/jobflow/issues/422)) transitioned from Pydantic v1 to Pydantic v2 recently. These two versions of Pydantic are not intercomparable. 

My assumption is that one of your dependencies is for an older version running on Pydantic v1. Can you please do the following:

1. Report what version of quacc and Jobflow you are using.
2. Update quacc, which will ensure you have the most up-to-date dependencies.

We can debug from there. Also, just as a note, I wish to emphasize that the MP workflows in quacc are merely for demonstration purposes only at the moment. The one true source of MP workflows is in [atomate2](https://github.com/materialsproject/atomate2/blob/main/src/atomate2/vasp/flows/mp.py). I make no guarantees of 1:1 compatibility at the moment, as the MP workflow in atomate2 was only just added the other day, and further validation is needed to ensure they match (for instance, see [this PR](https://github.com/Quantum-Accelerators/quacc/pull/1033)).

**Reply by JohnFengg** [reply]
Thanks very much for your quick reply @Andrew-S-Rosen !!!  I dont know how to get the version of quacc and jobflow package (sorry for that, I'm a beginner of python) :( , but I follow the quacc documentation to install the ase, quacc and jobflow again as the command:
       $pip install --no-cache-dir https://gitlab.com/ase/ase/-/archive/master/ase-master.zip
       $pip install quacc
       $pip install quacc[jobflow]
And all I got is "Requirement already satisfied:" stuff, does that means I already got the lastest version of the quacc and jobflow? The error seems still exists.
And Thanks again for your kind notice, I think the quacc package is amazing and suited for me as a beginner, especially when I am relying on the ase package to do some job with vasp, the pre-defined recipes in quacc are easier to use for me at this stage.
Appreciate for your time!

**Reply by Andrew-S-Rosen** [reply]
@JohnFengg:

Please run

```bash
pip uninstall quacc
pip install --no-cache-dir https://gitlab.com/ase/ase/-/archive/master/ase-master.zip
pip install --upgrade quacc[jobflow]
```

To get the versions:

```python
import jobflow as jf
import quacc

print(quacc.__version__)
print(jf.__version__)
```

Assuming they are the most recent releases on PyPI, please let me know if the error persists. Note that there are other issues with your script, but one thing at a time...

**Reply by JohnFengg** [reply]
Hello @Andrew-S-Rosen, I followed the command above, when I tried to import jobflow and quacc, it come the error:

> Traceback (most recent call last):
  File "/home/toussaint/high-throughput/model-run.py", line 3, in <module>
    from quacc.recipes.vasp.mp import mp_relax_flow
  File "/home/toussaint/opt/miniconda3/envs/quacc/lib/python3.10/site-packages/quacc/__init__.py", line 11, in <module>
    from quacc.settings import QuaccSettings
  File "/home/toussaint/opt/miniconda3/envs/quacc/lib/python3.10/site-packages/quacc/settings.py", line 20, in <module>
    import_module(wflow_engine)
  File "/home/toussaint/opt/miniconda3/envs/quacc/lib/python3.10/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "/home/toussaint/opt/miniconda3/envs/quacc/lib/python3.10/site-packages/jobflow/__init__.py", line 14, in <module>
    SETTINGS = JobflowSettings()
  File "/home/toussaint/opt/miniconda3/envs/quacc/lib/python3.10/site-packages/pydantic_settings/main.py", line 71, in __init__
    super().__init__(
  File "/home/toussaint/opt/miniconda3/envs/quacc/lib/python3.10/site-packages/pydantic/main.py", line 164, in __init__
    __pydantic_self__.__pydantic_validator__.validate_python(data, self_instance=__pydantic_self__)
pydantic_core._pydantic_core.ValidationError: 1 validation error for JobflowSettings
docs_store
  Extra inputs are not permitted [type=extra_forbidden, input_value={'type': 'MongoStore', 'h...'password': 'toussaint'}, input_type=CommentedMap]
    For further information visit https://errors.pydantic.dev/2.4/v/extra_forbidden

Could you please give some advise about how to fix that? Thanks very much! : )

**Reply by Andrew-S-Rosen** [reply]
@JohnFengg: I still need to know your versions to be able to assist you here. The error remains the same as before --- it's related to Pydantic. If you make a new Python environment and install `quacc[jobflow]`, this error will not appear.

I should note that this issue actually doesn't have anything directly to do with quacc. If you just do `import jobflow`, you're going to get the same error.

**Reply by Andrew-S-Rosen** [reply]
@JohnFengg: Ah, now I see the full traceback. The error is stated above: `pydantic_core._pydantic_core.ValidationError: 1 validation error for JobflowSettings`. In other words, your `~/.jobflow.yaml` file is not valid.

If you have further issues with importing jobflow, I recommend opening an issue on the jobflow repo.

**Reply by JohnFengg** [reply]
I tried to make a new Python environment and install quacc[jobflow], and when I import quacc and jobflow, I got same error as before. Maybe I should ask for help on jobflow repo.
by the way, my jobflow.yaml file is refered to the one in quacc docs, like:

```
docs_store: 
   type: MongoStore
   host: localhost
   port: 27017
   database: ase
   collection_name: ase
   username: toussaint
   password: toussaint
```

and I exported the path in `~/.bashrc`. Anyway, thanks for the helping, if there's progress I'll let you know.

**Reply by Andrew-S-Rosen** [reply]
@JohnFengg: Thanks. With the further details, I now know your issue. There should be two spaces preceeding `docs_store` so that it is properly indented in the YAML. Thank you for raising the issue. I have updated the docs.

https://github.com/Quantum-Accelerators/quacc/commit/cfec55d35c2f7ac13e5f539a3f2d084e9834afbf

**Reply by JohnFengg** [reply]
WoW!!! It works! Now I can succeefully import the jobflow and quacc! Thanks very much for your help !  
Could you please point the issue in my scripts (as you mentioned above).
I'm trying to do relax-job using vasp for hundreds of structures, I think the `mp_relax_flow` is sufficient and simple for me to deal with that, how do you think? For further improvement of my workflow, is that possible to apply the `custodian`  to the `mp_prerelax_job` or 
`mp_relax_flow` , or I have to create @job and @flow based on `custodian` . Forgive my confused logic, I'll apprecieate if I can get some similar example or some information such as notebooks....
Thanks again.

**Reply by JohnFengg** [reply]
Hello! @Andrew-S-Rosen  I have make some revise to my scripts:
```

from ase.db import connect
from quacc.recipes.vasp.mp import mp_relax_flow
from ase.io import read
import os



# Connect to the ASE database
database_path = f'{os.getcwd()}/perovskites.db'
db = connect(str(database_path))
atom_id=1
structure = db.get_atoms(id=atom_id)
workflow = mp_relax_flow(	
			atoms=structure,		
			prerelax_job_kwargs=None,
			relax_job_kwargs=None,
			)				
results = workflow["output"]
with open ('result.dat') as f:
	f.write(results)
db.close()

```

but I got the error like:

> Traceback (most recent call last):
>   File "/home/toussaint/high-throughput/test-run.py", line 14, in <module>
>     workflow = mp_relax_flow(
>   File "/home/toussaint/opt/miniconda3/envs/quacc/lib/python3.10/site-packages/quacc/recipes/vasp/mp.py", line 154, in mp_relax_flow
>     bandgap = prerelax_results["output"].get("bandgap", 0)
> TypeError: 'Job' object is not subscriptable
> 

could please give me some advise?

**Reply by Andrew-S-Rosen** [reply]
@JohnFengg: If you are looking to do a simple VASP relaxation, I would recommend trying [`quacc.recipes.vasp.core.relax_job`](https://quantum-accelerators.github.io/quacc/reference/quacc/recipes/vasp/core.html#quacc.recipes.vasp.core.relax_job).

That would look something like:

```python
import jobflow as jf
from quacc.recipes.vasp.core import relax_job
job = relax_job(structure)
response = jf.run_locally(job) # query `response` for the result
```

The reason for this is that, as noted in the docs, Jobflow does not support `_flow`-based recipes natively.

However, given the fact that you don't seem to be using FireWorks, what is your need for using Jobflow? If you are just looking to connect to a database, that can be done without Jobflow. You can run locally (`quacc set WORKFLOW_ENGINE local`) and have your database info in the `~/.quacc.yaml` as noted [here](https://quantum-accelerators.github.io/quacc/user/advanced/database.html). Then it would be as simple as:

```python
from quacc.recipes.vasp.core import relax_job
results = relax_job(structure)
```

**Reply by JohnFengg** [reply]
Hi! @Andrew-S-Rosen  Thanks for your advise! Actually, I'm trying to figure out the relationship between quacc, jobflow and fireworks... I want to run the jobs on HPC, it seems that the fireworks is necessary. Anyway, I have to do the thing step by step... 
Could you help me for the new error? I followed the above code:
```

from quacc.recipes.vasp.core import relax_job
import jobflow as jf
from ase.build import bulk

structure=bulk("Cu")
job = relax_job(structure)				
results = jf.run_locally(job)
print(results)


```
 And there comes the error about custodian, it seems that the INCAR was unreadable for custodian:

> 2023-10-10 12:47:32,672 INFO Started executing jobs locally
> 2023-10-10 12:47:32,673 INFO Starting job - relax_job (1242940d-446b-41b5-95df-cbe0ae55c4df)
> Error code was IERR= 5
> Error reading item
> /home/toussaint/opt/miniconda3/envs/quacc/lib/python3.10/site-packages/custodian/vasp/handlers.py:611: UserWarning: Looks like you made a typo in the INCAR. Please double-check it.
>   warnings.warn("Looks like you made a typo in the INCAR. Please double-check it.", UserWarning)
> ERROR:custodian.custodian:VaspErrorHandler
> ERROR:custodian.custodian:Unrecoverable error for handler: <custodian.vasp.handlers.VaspErrorHandler object at 0x7ff25658ffa0>
> Traceback (most recent call last):
>   File "/home/toussaint/opt/miniconda3/envs/quacc/lib/python3.10/site-packages/quacc/calculators/custodian/vasp.py", line 178, in <module>
>     run_custodian()
>   File "/home/toussaint/opt/miniconda3/envs/quacc/lib/python3.10/site-packages/quacc/calculators/custodian/vasp.py", line 174, in run_custodian
>     c.run()
>   File "/home/toussaint/opt/miniconda3/envs/quacc/lib/python3.10/site-packages/custodian/custodian.py", line 383, in run
>     self._run_job(job_n, job)
>   File "/home/toussaint/opt/miniconda3/envs/quacc/lib/python3.10/site-packages/custodian/custodian.py", line 527, in _run_job
>     raise NonRecoverableError(s, True, x["handler"])
> custodian.custodian.NonRecoverableError: Unrecoverable error for handler: <custodian.vasp.handlers.VaspErrorHandler object at 0x7ff25658ffa0>
> 2023-10-10 12:47:46,649 INFO relax_job failed with exception:
> Traceback (most recent call last):
>   File "/home/toussaint/opt/miniconda3/envs/quacc/lib/python3.10/site-packages/jobflow/managers/local.py", line 102, in _run_job
>     response = job.run(store=store)
>   File "/home/toussaint/opt/miniconda3/envs/quacc/lib/python3.10/site-packages/jobflow/core/job.py", line 583, in run
>     response = function(*self.function_args, **self.function_kwargs)
>   File "/home/toussaint/opt/miniconda3/envs/quacc/lib/python3.10/site-packages/quacc/recipes/vasp/core.py", line 143, in relax_job
>     return _base_job(
>   File "/home/toussaint/opt/miniconda3/envs/quacc/lib/python3.10/site-packages/quacc/recipes/vasp/core.py", line 255, in _base_job
>     atoms = run_calc(atoms, copy_files=copy_files)
>   File "/home/toussaint/opt/miniconda3/envs/quacc/lib/python3.10/site-packages/quacc/runners/calc.py", line 68, in run_calc
>     atoms.get_potential_energy()
>   File "/home/toussaint/opt/miniconda3/envs/quacc/lib/python3.10/site-packages/ase/atoms.py", line 730, in get_potential_energy
>     energy = self._calc.get_potential_energy(self)
>   File "/home/toussaint/opt/miniconda3/envs/quacc/lib/python3.10/site-packages/ase/calculators/abc.py", line 24, in get_potential_energy
>     return self.get_property(name, atoms)
>   File "/home/toussaint/opt/miniconda3/envs/quacc/lib/python3.10/site-packages/ase/calculators/calculator.py", line 502, in get_property
>     self.calculate(atoms, [name], system_changes)
>   File "/home/toussaint/opt/miniconda3/envs/quacc/lib/python3.10/site-packages/ase/calculators/vasp/vasp.py", line 326, in calculate
>     raise calculator.CalculationFailed(
> ase.calculators.calculator.CalculationFailed: vasp in . returned an error: 1
> 
> 2023-10-10 12:47:46,649 INFO Finished executing jobs locally
> {}

Does that means I miss some extra settings or something else? 
Thanks for your generous help!

**Reply by Andrew-S-Rosen** [reply]
There is only one way to find out why the error occurred, and that's to go to the directory and check the log files. If you share them here, I can tell you why. I can't diagnose the issue without seeing your INCAR. 

As for HPC, can you explain why you feel Jobflow and Fireworks is the only way to run quacc on HPC? That's not the case, and I'd like to understand where that misconception came from. All of the workflow engines let you run on HPC. In fact, I would generally recommend Parsl for that task given that it's the easiest to use. But yes, if you feel the need to use Jobflow, you must use fireworks to execute the tasks remotely.

**Reply by Andrew-S-Rosen** [reply]
I have an idea. It is likely that you're running with VASP < 6.4. Is this the case? By default, quacc writes out [EFERMI = MIDGAP](https://github.com/Quantum-Accelerators/quacc/blob/main/src/quacc/calculators/presets/vasp/BulkSet.yaml) to the INCAR since [VASP recommends](https://www.vasp.at/wiki/index.php/EFERMI) using that in all calculations. But it's only available in VASP 6.4+. You can use `calc_swaps = {"efermi": None}` to remove it.

**Reply by JohnFengg** [reply]
here is my INCAR file:

> INCAR created by Atomic Simulation Environment
>  ENAUG = 1360.000000
>  ENCUT = 680.000000
>  KSPACING = 0.220000
>  SIGMA = 0.050000
>  EDIFF = 1.00e-05
>  EDIFFG = -5.00e-02
>  ALGO = all
>  GGA = PS
>  PREC = accurate
>  EFERMI = midgap
>  IBRION = 2
>  ISIF = 3
>  ISMEAR = 0
>  LMAXMIX = 4
>  LORBIT = 11
>  NELM = 200
>  NSW = 99
>  LAECHG = .FALSE.
>  LASPH = .TRUE.
>  LCHARG = .TRUE.
>  LVTOT = .TRUE.
>  LWAVE = .FALSE.
>  LMIXTAU = .TRUE.
>  LREAL = .FALSE.
>  ISPIN = 2
>  MAGMOM = 1*1.0000

Oh I see, my vasp version is lower, so it's maybe the problem.
As for the choice of workflow engine, actually I dont have such thing like "I must use Jobflow and Fireworks" ( Perhaps just think it's popular, haha, sorry). And yeah, I'll follow your advise and learn about Parsl if it's easier for me.

**Reply by Andrew-S-Rosen** [reply]
If you look at the OUTCAR, you will almost certainly see that it complains about EFERMI = MIDGAP. So, in your case, I would suggest disabling it entirely and rerunning your workflow. That should resolve your issue.

I will likely try and add a new quacc setting called `VASP_MIN_VERSION` that the user can set. If `VASP_MIN_VERSION < 6.4`, then it won't ever add `EFERMI = MIDGAP`, such that you don't have to call `calc_swaps = {"efermi": None}` in every recipe.

Regarding workflow engine, of course you can feel free to use whatever you like. However, for a new user, you may find FireWorks to be a challenge unless you are familiar with MongoDB. Parsl is likely your best bet in my personal opinion, once you confirm things work locally.

EDIT: I opened PR https://github.com/Quantum-Accelerators/quacc/pull/1056 to add the `VASP_MIN_VERSION` feature.

**Reply by JohnFengg** [reply]
Yeah, the OUTCAR indeed reports the EFERMI error, however, when I add the `calc_swaps = {"efermi": None}` in relax_job(), the EFERMI stuff is still in INCAR, here is my scripts:
```

from quacc.recipes.vasp.core import relax_job
import jobflow as jf
from ase.build import bulk

structure=bulk("Cu")
job = relax_job(
                structure,
                calc_swaps = {"efermi": None},
)		
results = jf.run_locally(job)
print(results)
```

**Reply by Andrew-S-Rosen** [reply]
That seems like a problem. Let me get that patched and report back shortly.

**Reply by Andrew-S-Rosen** [reply]
@JohnFengg: It's getting a bit late here so I'll have to fix it tomorrow. I'll ping you when it's patched so you can update your quacc version.

The issue is that the VASP [INCAR "co-pilot" routine](https://github.com/Quantum-Accelerators/quacc/blob/cc303e7051586034d0093c355df90f8219a638a0/src/quacc/calculators/vasp.py#L315-L326), which tries to correct INCAR parameters on-the-fly, is changing "efermi" back to "midgap" to abide by the VASP manual. Instead, what we want to do is always respect `calc_swaps` and assume the user knows what they are doing.

In the meantime, you can do `quacc set VASP_INCAR_COPILOT False` in the command-line to disable the co-pilot routine so it won't override your changes automatically.

**Reply by JohnFengg** [reply]
OK, I see. (Get more knowledge about this program :P !) Cant imagine how difficult this work will be for me without your efficient help ! 
And the command seems a good choice temporarily for me to continute the job.
Good night!!!

**Reply by Andrew-S-Rosen** [reply]
@JohnFengg: Thanks for your patience. I have patched the issue that was preventing your `calc_swaps = {"efermi": None}` to register.

A new version of quacc will be minted on PyPI in approximately 10 mins, after which you can do:

```bash
pip uninstall quacc
pip install --upgrade quacc
```

**Reply by JohnFengg** [reply]
Got it ! @Andrew-S-Rosen  Thanks for your supporting. I'll try it later, and now I try to learn how to use the` parls ` and it's indeed easier to understand, thanks for the advise! Keep in touch :p.
